src/main/scala/mesosphere/marathon/state/TaskKiller.scala

```diff
@@ -0,0 +1,43 @@
+package mesosphere.marathon.state
+
+import javax.inject.Inject
+
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.tasks.TaskTracker
+import mesosphere.marathon.{ MarathonSchedulerService, UnknownAppException }
+
+import scala.concurrent.Future
+
+class TaskKiller @Inject() (
+    taskTracker: TaskTracker,
+    groupManager: GroupManager,
+    service: MarathonSchedulerService) {
+
+  def kill(appId: PathId,
+           findToKill: (Set[MarathonTask] => Set[MarathonTask]),
+           scale: Boolean): Future[Set[MarathonTask]] = {
+    import mesosphere.util.ThreadPoolContext.context
+
+    if (taskTracker.contains(appId)) {
+      val tasks = taskTracker.get(appId)
+      val toKill = findToKill(tasks)
+      if (scale) {
+        def scaleExisting(opt: Option[AppDefinition]): AppDefinition = opt
+          .map(app => app.copy(instances = app.instances - toKill.size))
+          .getOrElse(throw new UnknownAppException(appId))
+
+        val deploymentPlan = groupManager.updateApp(
+          appId, scaleExisting, Timestamp.now(), force = false, toKill = toKill)
+        deploymentPlan.map(_ => toKill)
```

Comment:
Good point. There are now 2 methods, killTasks and killTasksAndScale – the first one returns a `Future[Set[MarathonTask]]`, the other one a `Future[DeploymentPlan]`.